### ISSUE TITLE
Make rescans faster

### ISSUE THREAD

**Issue by jamesob** [bitcoin/bitcoin#23727]
This issue aggregates ideas for making wallet rescans substantially faster.

Using compact block filters, when available, could substantially speed up wallet rescans although it is not clear to me how this might work for non-descriptor-based (legacy) wallets.

Currently, #20664 proposes to add an RPC that will use compact block filters to scan for blocks relevant to a given script descriptor. This gives a listing of blocks that correspond to the descriptor, but the user then has to do something with that list of blocks to find transactions relevant to their wallet. 

It would be nice to integrate scanblocks' use of block filters opaquely into `rescanblockchain` (wallet rescanning) so that if block filters are available, the faster method of consulting them could be used instead of scanning the entire chain, or forcing the user to do something more roundabout with the output of `scanblocks`.

From @sipa: 
> regarding faster rescanning, i think there is some lower-hanging fruit, like (a) rescanning all wallets at startup in parallel rather than one by one and (b) building a set of scriptPubKeys to look for and using that for (legacy wallet) rescanning, instead of calling IsMine for each (especially with achow101's migrate-to-descriptor code, i think there should be a way of computing that set for legacy wallets too)

### Possible changes

- [ ] integrate compact block filter usage into `rescanblockchain` to allow faster rescans when the compact block index is available, otherwise falling back to the standard scanning method. A first approach could possibly use filters only when dealing with a non-legacy descriptor wallet
    - Currently proposed in https://github.com/bitcoin/bitcoin/pull/23549
- [ ] allow pruned nodes with the block filter index to rescan quickly by getting a list of relevant blocks to the wallet, requesting the download of blocks, and then parsing the blocks for relevant transactions
- [ ] rescanning all wallets at startup in parallel (instead of one-by-one): https://github.com/bitcoin/bitcoin/issues/11756
- [ ] building a set of scriptPubKeys to look for when using a legacy wallet based on @achow101's migrate-to-descriptor code

**Comment by ryanofsky** [comment]
>[ ]  rescanning all wallets at startup in parallel (instead of one-by-one)

This is #11756, and I'm working on it now, but a little indirectly. Specifically I'm:

1. Porting `BaseIndex::ThreadSync` and other index code to sync with `interfaces::Chain` instead of calling validation code directly. The motivation is to let indexes be used with [#10102](https://github.com/bitcoin/bitcoin/issues/10102) and run in different processes, but this also adds a new chain sync feature that can be used by wallets.
2. Introducing a block reader class to help with parallel rescans and pass same block data to multiple indexes.
3. Going to rebase [#15719](https://github.com/bitcoin/bitcoin/issues/15719) on top of this rescan code so wallets can use the same chain sync interface as the indexes

There's some uncertainty and performance tradeoffs in step 2. Right now my block reader implementation just reads 2 blocks at a time so previous block can be processed while next block is read. But it might be better to let indexes and wallets read blocks more freely without trying to schedule things.

**Comment by MarcoFalke** [comment]
> although it is not clear to me how this might work for non-descriptor-based (legacy) wallets.

Something like https://github.com/bitcoin/bitcoin/pull/15845#issuecomment-625768563 should do that

**Comment by MarcoFalke** [comment]
See also #19116

**Comment by Administratordes** [comment]
Hello GoodDay if you need any help or assistance regarding to any issue at all please don’t hesitate to contact the LiveChat Support https://direct.lc.chat/14065185/

**Comment by Administratordes** [comment]
Hello GoodDay if you need any help or assistance regarding to any issue at all please don’t hesitate to contact the LiveChat Support https://direct.lc.chat/14065185/

**Comment by Sjors** [comment]
PR description could now point to #23549

**Comment by jamesob** [comment]
Worth tagging https://github.com/bitcoin/bitcoin/pull/23549 here too.

**Comment by Sjors** [comment]
As well as #15845!

**Comment by achow101** [comment]
#25957 for fast rescan with filters was merged.

**Comment by MarcoFalke** [comment]
Can this be closed, given that the remaining thing #https://github.com/bitcoin/bitcoin/issues/11756 is already a separate issue?

**Comment by Sjors** [comment]
Blazing fast as of #25957, so I think we can close this.
